gensim/models/poincare.py

```diff
@@ -0,0 +1,828 @@
+from gensim.models.keyedvectors import KeyedVectors, Vocab
+from gensim.models.word2vec import Word2Vec
+
+logger = logging.getLogger(__name__)
+
+
+class PoincareModel(utils.SaveLoad):
+    """Class for training, using and evaluating Poincare Embeddings.
+
+    The model can be stored/loaded via its `save()` and `load()` methods, or stored/loaded in the word2vec
+    format via `wv.save_word2vec_format()` and `KeyedVectors.load_word2vec_format()`.
+
+    Note that training cannot be resumed from a model loaded via `load_word2vec_format`, if you wish to train further,
+    use `save()` and `load()` methods instead.
+
+    """
+    def __init__(
+        self, train_data, size=50, alpha=0.1, negative=10, iter=50,
+        workers=1, epsilon=1e-5, burn_in=10, burn_in_alpha=0.01, seed=0):
+        """Initialize and train a Poincare embedding model from an iterable of transitive closure relations.
+
+        Parameters
+        ----------
+        train_data : iterable of (str, str)
+            Iterable of relations, e.g. a list of tuples, or a PoincareRelations instance streaming from a file.
+        size : int, optional
+            Number of dimensions of the trained model.
+        alpha : float, optional
+            Learning rate for training.
+        negative : int, optional
+            Number of negative samples to use.
+        iter : int, optional
+            Number of iterations (epochs) over the corpus.
+        workers : int, optional
+            Number of threads to use for training the model.
+        epsilon : float, optional
+            Constant used for clipping embeddings below a norm of one.
+        burn_in : int, optional
+            Number of epochs to use for burn-in initialization (0 means no burn-in).
+        burn_in_alpha : float, optional
+            learning rate for burn-in initialization, ignored if `burn_in` is 0.
+        seed : int, optional
+            Seed for random to ensure reproducibility.
+
+        Examples
+        --------
+        Initialize a model from a list::
+
+        >>> from gensim.models.poincare import PoincareModel
+        >>> relations = [('kangaroo', 'marsupial'), ('kangaroo', 'mammal'), ('gib', 'cat')]
+        >>> model = PoincareModel(relations, negative=2)
+
+        Initialize a model from a file containing one relation per line::
+
+        >>> from gensim.models.poincare import PoincareModel, PoincareRelations
+        >>> file_path = 'gensim/test/test_data/poincare_hypernyms.tsv'
+        >>> model = PoincareModel(PoincareRelations(file_path), negative=2)
+
+        See `PoincareRelations` for more options.
+
+        """
+        self.train_data = train_data
+        self.wv = KeyedVectors()
+        self.size = size
+        self.train_alpha = alpha  # Learning rate for training
+        self.burn_in_alpha = burn_in_alpha  # Learning rate for burn-in
+        self.alpha = alpha  # Current learning rate
+        self.negative = negative
+        self.iter = iter
```

Comment:
`iter` collides with a built-in function. `iterations` would be better, IMO.

Reply:
Was following word2vec API which uses `iter`. Renamed to `epochs` now, shorter and more specific/clear IMO.